Add numba sigmoid and GammaLn implementation

aseyboldt opened the pull request:
Here are a few important guidelines and requirements to check before your PR can be merged:
+ [x] There is an informative high-level description of the changes.
+ [ ] The description and/or commit message(s) references the relevant GitHub issue(s).
+ [ ] [`pre-commit`](https://pre-commit.com/#installation) is installed and [set up](https://pre-commit.com/#3-install-the-git-hook-scripts).
+ [ ] The commit messages follow [these guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [ ] The commits correspond to [_relevant logical changes_](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes), and there are **no commits that fix changes introduced by other commits in the same branch/BR**.
+ [ ] There are tests covering the changes introduced in the PR.

Currently unless numba-scipy is installed (and it isn't a dependency) compiling pymc models that use sigmoid fails, eg

```python
with pm.Model() as model:
    pm.Uniform("a", lower=0, upper=1)
```

Sigmoid is so simple however, that I'm not sure we really need to call into scipy for it, we could just provide a custom implementation directly as in this PR?

brandonwillard commented on aesara/link/numba/dispatch/scalar.py:
```diff
@@ -241,3 +242,15 @@ def inv(x):
     return inv
+
+
+# Override the more general ScalarOp implementation
+# to avoid using scipy.special in numba and make the
+# implementation transparent for numba
```
```suggestion
```
This doesn't really need an explanation, especially since the `numba-scipy` logic is only a fallback for missing `ScalarOp` implementations.

aseyboldt commented:
@brandonwillard I removed the comment you mentioned, and also added an implementation for gammaln.
The tests in test_numba are still valid, but now we don't require numba_scipy for anything in this file. It seems it gets installed when running the tests, but is not a requirement in conda-forge.
Do we actually want to add this as a dependency?

brandonwillard commented:
> Do we actually want to add this as a dependency?

I think I hesitated to do that initially because `numba-scipy` put restrictions on the SciPy version.

aseyboldt commented:
I see.
I'm not really sure if we need this PR then.
A lot of models would work without numba_scipy with those two functions (and maybe a couple more?). Using custom implementations *might* also be faster, because the compiler actually sees the function implementation and can inline.
But on the other hand there might be more `scipy.special` functions that come up from time to time and are tricky to get anywhere else...

brandonwillard commented:
> I'm not really sure if we need this PR then.

I was only explaining why I think we didn't include `numba-scipy` as a dependency in the conda-forge recipe.

Regarding the test environment/dependencies: we should still have a test for cooperability with `numba-scipy`, but we can simultaneously test the `ScalarOp` implementations in this PR.

I think these `ScalarOp` implementations will be used before the `numba-scipy` approach when/if there are entries present for them in the dispatch function, so there shouldn't be any problems with testing your changes; otherwise, we might need to add a test case that uses a `scipy.special` function with no `ScalarOp` dispatch, in order to get coverage for the `numba-scipy` case.

> But on the other hand there might be more `scipy.special` functions that come up from time to time and are tricky to get anywhere else...

Yeah, but we're only using it now because it provides quick coverage of multiple unimplemented `ScalarOp`s.  As you said, if we can produce our own implementations, then we're likely much better off.

aseyboldt commented:
I now use object mode if numba_scipy can't be imported. Maybe an approach like this is a bit safer?
To test it properly I guess we need to add a config value as well, so that we can just install numba_scipy in the test env, but it in one test matrix item?

I also changed which apply node is passed to `numba_funcify` in elemwise. I don't think it makes sense to pass the elemwise node to the funcify function of the scalar_op, does it?
https://github.com/aesara-devs/aesara/pull/1062/files#diff-1974aff2f50ca4ebb89282f584d28919e83c38ab38dba1cb6791e07b26ce0403R428

aseyboldt commented on aesara/link/numba/dispatch/elemwise.py:
```diff
@@ -424,8 +425,19 @@ def axis_apply_fn(x):
 
 @numba_funcify.register(Elemwise)
 def numba_funcify_Elemwise(op, node, **kwargs):
+    # A FunctionGraph isn't a Node in the type hierarchy, but
```
I'm not really sure what to do here.
Previously, we just passed the node of the surrounding op, but I think this clearly isn't right, and leads to errors if we want to generate a object mode fallback for parts of the elemwise code.

So I thought we can just create a new node using `op.scalar_op.make_node(*scalar_inputs)`, but this doesn't always work (maybe actually due to a bug). Elemwise will sometimes throw errors because it thinks something in the graph isn't a scalar op.

So my conclusion was that we can just pass in the fgraph, which also has input and output arguments just as a node, and I think pretty much functions as the node conceptually. But this isn't exactly clean.

Maybe we need to figure out what's going wrong with the make_node call?

aseyboldt replied:
^ @brandonwillard if you have an idea about what to do here, that would be great.

brandonwillard replied:
> I'm not really sure what to do here.
> Previously, we just passed the node of the surrounding op, but I think this clearly isn't right, and leads to errors if we want to generate a object mode fallback for parts of the elemwise code.

Are you trying to make the scalar `Op` call in `numba_funcify_Elemwise` return an object-mode fallback implementation?

brandonwillard reviewed (commented):
Before diving into any details regarding `numba-scipy` and/or Aesara interface refactoring, we should probably go forward with the `ScalarOp` dispatch additions first and address the rest in a subsequent PR.

brandonwillard commented:
> I also changed which apply node is passed to `numba_funcify` in elemwise. I don't think it makes sense to pass the elemwise node to the funcify function of the scalar_op, does it?
> https://github.com/aesara-devs/aesara/pull/1062/files#diff-1974aff2f50ca4ebb89282f584d28919e83c38ab38dba1cb6791e07b26ce0403R428

I think that was done as a convenience for `Type`-related information needed by `numba_funcify_ScalarOp`

aseyboldt commented:
Split into #1077 and #1078
